homeassistant/core.py

```diff
@@ -408,6 +433,8 @@ def async_fire(self, event_type: str, event_data=None,
         for func in listeners:
```

Comment:
Well, 'queue' them for immediate execution on the thread-pool. Given Python thread-switching that means that it was entirely likely that jobs queued would actually start running before the next one was queued. Now that a bunch of the listeners will actually run it if it is a callback avoids yet another trip through the event loop.

The one nice thing I could see about immediately executing callback's is that a callback expecting the state to be a certain way will actually have it that way when its run. I don't think there's any problem with a listener 'skipping the line' since the line could be totally unrelated to the event that is actually occurring.